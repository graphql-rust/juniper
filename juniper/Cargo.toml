[package]
name = "juniper"
version = "0.11.1"
authors = [
    "Magnus Hallin <mhallin@fastmail.com>",
    "Christoph Herzog <chris@theduke.at>",
]
description = "GraphQL server library"
license = "BSD-2-Clause"
documentation = "https://docs.rs/juniper"
repository = "https://github.com/graphql-rust/juniper"
readme = "../README.md"
keywords = ["graphql", "server", "web", "rocket"]
categories = ["web-programming"]

[badges]
travis-ci = { repository = "graphql-rust/juniper" }

[[bench]]
name = "bench"
harness = false
path = "benches/bench.rs"

[features]
nightly = []
expose-test-schema = []
default = [
    "chrono",
    "url",
    "uuid",
]

[dependencies]
juniper_codegen = { version = "0.11.1", path = "../juniper_codegen"  }

fnv = "1.0.3"
indexmap = { version = "1.0.0", features = ["serde-1"] }
serde = { version = "1.0.8" }
serde_derive = { version = "1.0.2" }

chrono = { version = "0.4.0", optional = true }
serde_json = { version="1.0.2", optional = true }
url = { version = "1.5.1", optional = true }
uuid = { version = "0.7", optional = true }

[dev-dependencies]
bencher = "0.1.2"
serde_json = { version = "1.0.2" }
