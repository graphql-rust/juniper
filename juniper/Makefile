###############################
# Common defaults/definitions #
###############################

# Checks two given strings for equality.
eq = $(if $(or $(1),$(2)),$(and $(findstring $(1),$(2)),\
                                $(findstring $(2),$(1))),1)

# Multiplatform prefix of `sed -i` commands.
sed-i = sed -i$(if $(call eq,$(shell uname -s),Darwin), '',)




######################
# Project parameters #
######################

GRAPHIQL_VER ?= $(strip \
	$(shell grep -m1 '"graphiql": "' package.json | cut -d '"' -f4))
GRAPHQL_PLAYGROUND_VER ?= $(strip \
	$(shell grep -m1 '"graphql-playground-react": "' package.json | cut -d '"' -f4))




############
# Commands #
############

# Download and prepare actual version of GraphiQL static files, used for
# integrating it.
#
# Usage:
#	make graphiql

graphiql:
	curl -fL -o src/http/graphiql.html \
		https://raw.githubusercontent.com/graphql/graphiql/graphiql%40$(GRAPHIQL_VER)/examples/graphiql-cdn/index.html
	$(sed-i) 's|unpkg.com/graphiql/|unpkg.com/graphiql@$(GRAPHIQL_VER)/|g' \
		src/http/graphiql.html
	$(sed-i) 's|react.development.js|react.production.min.js|g' \
		src/http/graphiql.html
	$(sed-i) 's|react-dom.development.js|react-dom.production.min.js|g' \
		src/http/graphiql.html
	$(sed-i) "s|'https://swapi-graphql.netlify.app/.netlify/functions/index'|JUNIPER_URL|g" \
		src/http/graphiql.html
	$(sed-i) "s|url: JUNIPER_URL,|url: JUNIPER_URL,\n        subscriptionUrl: normalizeSubscriptionEndpoint(JUNIPER_URL, JUNIPER_SUBSCRIPTIONS_URL)|" \
		src/http/graphiql.html
	$(sed-i) 's|<script>|<script>\n<!-- inject -->|' \
		src/http/graphiql.html
	$(sed-i) '/X-Example-Header/d' \
		src/http/graphiql.html


# Download and prepare actual version of GraphQL Playground static files, used
# for integrating it.
#
# Usage:
#	make graphql-playground

graphql-playground:
	curl -fL -o src/http/playground.html \
		https://raw.githubusercontent.com/graphql/graphql-playground/graphql-playground-react%40$(GRAPHQL_PLAYGROUND_VER)/packages/graphql-playground-html/withAnimation.html
	$(sed-i) 's|cdn.jsdelivr.net/npm/graphql-playground-react/|cdn.jsdelivr.net/npm/graphql-playground-react@$(GRAPHQL_PLAYGROUND_VER)/|g' \
		src/http/playground.html
	$(sed-i) "s|// you can add more options here|endpoint: 'JUNIPER_URL', subscriptionEndpoint: 'JUNIPER_SUBSCRIPTIONS_URL'|" \
		src/http/playground.html




##################
# .PHONY section #
##################

.PHONY: graphiql graphql-playground
