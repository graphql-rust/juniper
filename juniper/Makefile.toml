# This is needed as the release config is at a different path than the top-level
# release config.
[env]
CARGO_MAKE_CARGO_ALL_FEATURES = ""

[tasks.release-INTERNAL]
args = ["release", "--config", "${CARGO_MAKE_WORKING_DIRECTORY}/release.toml", "${RELEASE_LEVEL}"]

[tasks.release-dry-run-INTERNAL]
args = ["release", "--config", "${CARGO_MAKE_WORKING_DIRECTORY}/release.toml", "--dry-run", "${RELEASE_LEVEL}"]

[tasks.release-local-test-INTERNAL]
args = ["release", "--config", "${CARGO_MAKE_WORKING_DIRECTORY}/release.toml", "--no-confirm", "--skip-publish", "--skip-push", "--skip-tag", "${RELEASE_LEVEL}"]

[tasks.test]
args = ["test", "--all-features"]
[tasks.test-custom]
args = ["test", "--all-features"]
[tasks.test-flow]
args = ["test", "--all-features"]
[tasks.test-multi-flow-phase]
args = ["test", "--all-features"]
[tasks.test-thread-safe]
args = ["test", "--all-features"]
[tasks.test-verbose]
args = ["test", "--all-features"]
[tasks.test-with-args]
args = ["test", "--all-features"]
[tasks.ci-coverage-flow]
args = ["test", "--all-features"]
