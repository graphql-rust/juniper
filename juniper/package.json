{
  "private": true,
  "scripts": {
    "postinstall": "make graphiql graphql-playground"
  },
  "dependencies": {
    "graphiql": "3.1.0",
    "graphql-playground-react": "1.7.28"
  }
}
