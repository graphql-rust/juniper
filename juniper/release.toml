no-dev-version = true
pre-release-commit-message = "Release {{crate_name}} {{version}}"
pro-release-commit-message = "Bump {{crate_name}} version to {{next_version}}"
tag-message = "Release {{crate_name}} {{version}}"
upload-doc = false
pre-release-replacements = [
  # Juniper's changelog
  {file="CHANGELOG.md", search="# master", replace="# master\n\n- No changes yet\n\n# [[{{version}}] {{date}}](https://github.com/graphql-rust/juniper/releases/tag/{{crate_name}}-{{version}})"},
  {file="src/lib.rs", search="docs.rs/juniper/[a-z0-9\\.-]+", replace="docs.rs/juniper/{{version}}"},
  # docs
  {file="../docs/book/content/quickstart.md", search="juniper = \"[^\"]+\"", replace="juniper = \"{{version}}\""},
  # codegen
  {file="../juniper_codegen/Cargo.toml", search="juniper = \\{ version = \"[^\"]+\"", replace="juniper = { version = \"{{version}}\""},
  # Tests.
  {file="../integration_tests/juniper_tests/Cargo.toml", search="juniper = \\{ version = \"[^\"]+\"", replace="juniper = { version = \"{{version}}\""},
  # Hyper
  {file="../juniper_hyper/Cargo.toml", search="juniper = \\{ version = \"[^\"]+\"", replace="juniper = { version = \"{{version}}\""},
  {file="../juniper_hyper/Cargo.toml", search="\\[dev-dependencies\\.juniper\\]\nversion = \"[^\"]+\"", replace="[dev-dependencies.juniper]\nversion = \"{{version}}\""},
  # Iron
  {file="../juniper_iron/Cargo.toml", search="juniper = \\{ version = \"[^\"]+\"", replace="juniper = { version = \"{{version}}\""},
  {file="../juniper_iron/Cargo.toml", search="\\[dev-dependencies\\.juniper\\]\nversion = \"[^\"]+\"", replace="[dev-dependencies.juniper]\nversion = \"{{version}}\""},
  # Rocket
  {file="../juniper_rocket/Cargo.toml", search="juniper = \\{ version = \"[^\"]+\"", replace="juniper = { version = \"{{version}}\""},
  {file="../juniper_rocket/Cargo.toml", search="\\[dev-dependencies\\.juniper\\]\nversion = \"[^\"]+\"", replace="[dev-dependencies.juniper]\nversion = \"{{version}}\""},
  # Rocket Async
  {file="../juniper_rocket_async/Cargo.toml", search="juniper = \\{ version = \"[^\"]+\"", replace="juniper = { version = \"{{version}}\""},
  {file="../juniper_rocket_async/Cargo.toml", search="\\[dev-dependencies\\.juniper\\]\nversion = \"[^\"]+\"", replace="[dev-dependencies.juniper]\nversion = \"{{version}}\""},
  # Warp
  {file="../juniper_warp/Cargo.toml", search="juniper = \\{ version = \"[^\"]+\"", replace="juniper = { version = \"{{version}}\""},
  {file="../juniper_warp/Cargo.toml", search="\\[dev-dependencies\\.juniper\\]\nversion = \"[^\"]+\"", replace="[dev-dependencies.juniper]\nversion = \"{{version}}\""},
  # Subscriptions
  {file="../juniper_subscriptions/Cargo.toml", search="juniper = \\{ version = \"[^\"]+\"", replace="juniper = { version = \"{{version}}\""},
  # GraphQL-WS
  {file="../juniper_graphql_ws/Cargo.toml", search="juniper = \\{ version = \"[^\"]+\"", replace="juniper = { version = \"{{version}}\""},
  # Actix-Web
  {file="../juniper_actix/Cargo.toml", search="juniper = \\{ version = \"[^\"]+\"", replace="juniper = { version = \"{{version}}\""},
  {file="../juniper_actix/Cargo.toml", search="\\[dev-dependencies\\.juniper\\]\nversion = \"[^\"]+\"", replace="[dev-dependencies.juniper]\nversion = \"{{version}}\""},
]
