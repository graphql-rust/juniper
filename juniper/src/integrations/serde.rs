use indexmap::IndexMap;
use serde::ser::SerializeMap;
use serde::{de, ser};

use std::fmt;

use ast::InputValue;
use executor::ExecutionError;
use parser::{ParseError, SourcePosition, Spanning};
use validation::RuleError;
use {GraphQLError, Object, ScalarValue, Value};

#[derive(Serialize)]
struct SerializeHelper {
    message: &'static str,
}

impl<T> ser::Serialize for ExecutionError<T>
where
    T: ScalarValue,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: ser::Serializer,
    {
        let mut map = serializer.serialize_map(Some(4))?;

        map.serialize_key("message")?;
        map.serialize_value(self.error().message())?;

        let locations = vec![self.location()];
        map.serialize_key("locations")?;
        map.serialize_value(&locations)?;

        map.serialize_key("path")?;
        map.serialize_value(self.path())?;

        if !self.error().extensions().is_null() {
            map.serialize_key("extensions")?;
            map.serialize_value(self.error().extensions())?;
        }

        map.end()
    }
}

impl<'a> ser::Serialize for GraphQLError<'a> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: ser::Serializer,
    {
        match *self {
            GraphQLError::ParseError(ref err) => vec![err].serialize(serializer),
            GraphQLError::ValidationError(ref errs) => errs.serialize(serializer),
            GraphQLError::NoOperationProvided => [SerializeHelper {
                message: "Must provide an operation",
            }]
                .serialize(serializer),
            GraphQLError::MultipleOperationsProvided => [SerializeHelper {
                message: "Must provide operation name \
                          if query contains multiple operations",
            }]
                .serialize(serializer),
            GraphQLError::UnknownOperationName => [SerializeHelper {
                message: "Unknown operation",
            }]
                .serialize(serializer),
        }
    }
}

impl<'de, S> de::Deserialize<'de> for InputValue<S>
where
    S: ScalarValue,
{
    fn deserialize<D>(deserializer: D) -> Result<InputValue<S>, D::Error>
    where
        D: de::Deserializer<'de>,
    {
        struct InputValueVisitor<S: ScalarValue>(S::Visitor);

        impl<S: ScalarValue> Default for InputValueVisitor<S> {
            fn default() -> Self {
                InputValueVisitor(S::Visitor::default())
            }
        }

        impl<'de, S> de::Visitor<'de> for InputValueVisitor<S>
        where
            S: ScalarValue,
        {
            type Value = InputValue<S>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("a valid input value")
            }

            fn visit_bool<E>(self, value: bool) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_bool(value).map(InputValue::Scalar)
            }

            fn visit_i8<E>(self, value: i8) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_i8(value).map(InputValue::Scalar)
            }

            fn visit_i16<E>(self, value: i16) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_i16(value).map(InputValue::Scalar)
            }

            fn visit_i32<E>(self, value: i32) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_i32(value).map(InputValue::Scalar)
            }

            fn visit_i64<E>(self, value: i64) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_i64(value).map(InputValue::Scalar)
            }

            serde_if_integer128! {
                fn visit_i128<E>(self, value: i128) -> Result<InputValue<S>, E>
                where
                    E: de::Error,
                {
                    self.0.visit_i128(value).map(InputValue::Scalar)
                }
            }

            fn visit_u8<E>(self, value: u8) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_u8(value).map(InputValue::Scalar)
            }

            fn visit_u16<E>(self, value: u16) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_u16(value).map(InputValue::Scalar)
            }

            fn visit_u32<E>(self, value: u32) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_u32(value).map(InputValue::Scalar)
            }

            fn visit_u64<E>(self, value: u64) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_u64(value).map(InputValue::Scalar)
            }

            serde_if_integer128!{
                fn visit_u128<E>(self, value: u128) -> Result<InputValue<S>, E>
                where
                    E: de::Error,
                {
                    self.0.visit_u128(value).map(InputValue::Scalar)
                }
            }

            fn visit_f32<E>(self, value: f32) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_f32(value).map(InputValue::Scalar)
            }

            fn visit_f64<E>(self, value: f64) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_f64(value).map(InputValue::Scalar)
            }

            fn visit_char<E>(self, value: char) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_char(value).map(InputValue::Scalar)
            }

            fn visit_str<E>(self, value: &str) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_str(value).map(InputValue::Scalar)
            }

            fn visit_string<E>(self, value: String) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_string(value).map(InputValue::Scalar)
            }

            fn visit_bytes<E>(self, bytes: &[u8]) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_bytes(bytes).map(InputValue::Scalar)
            }

            fn visit_byte_buf<E>(self, bytes: Vec<u8>) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                self.0.visit_byte_buf(bytes).map(InputValue::Scalar)
            }

            fn visit_none<E>(self) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                Ok(InputValue::null())
            }

            fn visit_unit<E>(self) -> Result<InputValue<S>, E>
            where
                E: de::Error,
            {
                Ok(InputValue::null())
            }

            fn visit_seq<V>(self, mut visitor: V) -> Result<InputValue<S>, V::Error>
            where
                V: de::SeqAccess<'de>,
            {
                let mut values = Vec::new();

                while let Some(el) = visitor.next_element()? {
                    values.push(el);
                }

                Ok(InputValue::list(values))
            }

            fn visit_map<V>(self, mut visitor: V) -> Result<InputValue<S>, V::Error>
            where
                V: de::MapAccess<'de>,
            {
                let mut object = IndexMap::<String, InputValue<S>>::with_capacity(
                    visitor.size_hint().unwrap_or(0),
                );

                while let Some((key, value)) = visitor.next_entry()? {
                    object.insert(key, value);
                }

                Ok(InputValue::object(object))
            }
        }

        deserializer.deserialize_any(InputValueVisitor::default())
    }
}

impl<T> ser::Serialize for InputValue<T>
where
    T: ScalarValue,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: ser::Serializer,
    {
        match *self {
            InputValue::Null | InputValue::Variable(_) => serializer.serialize_unit(),
            InputValue::Scalar(ref s) => s.serialize(serializer),
            InputValue::Enum(ref v) => serializer.serialize_str(v),
            InputValue::List(ref v) => v
                .iter()
                .map(|x| x.item.clone())
                .collect::<Vec<_>>()
                .serialize(serializer),
            InputValue::Object(ref v) => v
                .iter()
                .map(|&(ref k, ref v)| (k.item.clone(), v.item.clone()))
                .collect::<IndexMap<_, _>>()
                .serialize(serializer),
        }
    }
}

impl ser::Serialize for RuleError {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: ser::Serializer,
    {
        let mut map = serializer.serialize_map(Some(2))?;

        map.serialize_key("message")?;
        map.serialize_value(self.message())?;

        map.serialize_key("locations")?;
        map.serialize_value(self.locations())?;

        map.end()
    }
}

impl ser::Serialize for SourcePosition {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: ser::Serializer,
    {
        let mut map = serializer.serialize_map(Some(2))?;

        let line = self.line() + 1;
        map.serialize_key("line")?;
        map.serialize_value(&line)?;

        let column = self.column() + 1;
        map.serialize_key("column")?;
        map.serialize_value(&column)?;

        map.end()
    }
}

impl<'a> ser::Serialize for Spanning<ParseError<'a>> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: ser::Serializer,
    {
        let mut map = serializer.serialize_map(Some(2))?;

        let message = format!("{}", self.item);
        map.serialize_key("message")?;
        map.serialize_value(&message)?;

        let mut location = IndexMap::new();
        location.insert("line".to_owned(), self.start.line() + 1);
        location.insert("column".to_owned(), self.start.column() + 1);

        let locations = vec![location];

        map.serialize_key("locations")?;
        map.serialize_value(&locations)?;

        map.end()
    }
}

impl<T> ser::Serialize for Object<T>
where
    T: ser::Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: ser::Serializer,
    {
        let mut map = serializer.serialize_map(Some(self.field_count()))?;

        for &(ref f, ref v) in self.iter() {
            map.serialize_key(f)?;
            map.serialize_value(v)?;
        }

        map.end()
    }
}

impl<T> ser::Serialize for Value<T>
where
    T: ser::Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: ser::Serializer,
    {
        match *self {
            Value::Null => serializer.serialize_unit(),
            Value::Scalar(ref s) => s.serialize(serializer),
            Value::List(ref v) => v.serialize(serializer),
            Value::Object(ref v) => v.serialize(serializer),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{ExecutionError, GraphQLError};
    use ast::InputValue;
    use serde_json::from_str;
    use serde_json::to_string;
    use value::{DefaultScalarValue, Object};
    use {FieldError, Value};

    #[test]
    fn int() {
        assert_eq!(
            from_str::<InputValue<DefaultScalarValue>>("1235").unwrap(),
            InputValue::scalar(1235)
        );
    }

    #[test]
    fn float() {
        assert_eq!(
            from_str::<InputValue<DefaultScalarValue>>("2.0").unwrap(),
            InputValue::scalar(2.0)
        );
        // large value without a decimal part is also float
        assert_eq!(
            from_str::<InputValue<DefaultScalarValue>>("123567890123").unwrap(),
            InputValue::scalar(123567890123.0)
        );
    }

    #[test]
    fn errors() {
        assert_eq!(
            to_string(&GraphQLError::UnknownOperationName).unwrap(),
            r#"[{"message":"Unknown operation"}]"#
        );
    }

    #[test]
    fn error_extensions() {
        let mut obj: Object<DefaultScalarValue> = Object::with_capacity(1);
        obj.add_field("foo".to_string(), Value::scalar("bar"));
        assert_eq!(
            to_string(&ExecutionError::at_origin(FieldError::new(
                "foo error",
                Value::Object(obj),
            ))).unwrap(),
            r#"{"message":"foo error","locations":[{"line":1,"column":1}],"path":[],"extensions":{"foo":"bar"}}"#
        );
    }
}
