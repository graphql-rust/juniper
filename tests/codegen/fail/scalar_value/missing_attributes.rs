use derive_more::{Display, From, TryInto};
use juniper::ScalarValue;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize, Display, From, PartialEq, ScalarValue, Serialize, TryInto)]
pub enum DefaultScalarValue {
    Int(i32),
    #[value(to_float)]
    Float(f64),
    #[value(as_str, to_string)]
    String(String),
    #[value(to_bool)]
    Boolean(bool),
}

fn main() {}
