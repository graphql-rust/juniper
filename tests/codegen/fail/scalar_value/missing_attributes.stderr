error: GraphQL built-in scalars missing `#[value(as_int, as_float)]` attributes. In case you are sure that it's ok, use `#[value(allow_missing_attributes)]` to suppress this error.
 --> fail/scalar_value/missing_attributes.rs:4:1
  |
4 | pub enum DefaultScalarValue {
  | ^^^
